### PR TITLE
Order Item Data Store

### PR DESCRIPTION
**Pull request by emdashcodes**
This PR implements data stores for all order item types. There is an abstract base class for order item stores, and then smaller stores that handle each specific set of data. This also updates the main data classes to use the new set_prop and get_prop methods.

#### Testing:

Unit Tests: All tests pass. `order/crud` tests have some line item tests specifically.

Cart/Ordering:
* Add a few different product types to your cart - enable things like shipping fees or taxes.
* Checkout and verify that info shows correctly on the order summary page (product name, cost, tax info, etc).

Admin:
* Verify that info shows correctly for the order line items in wp-admin.
* Add new fees, shipping costs, or products to the order. Save and verify everything works and displays correctly.

### CODE REVIEW

File: includes/class-wc-order-item-fee.php [hunk 1]
```diff
@@ -196,14 +108,18 @@ public function set_taxes( $raw_tax_data ) {
...
 	 * Get order item name.
+	 *
+	 * @param  string $context
 	 * @return string
 	 */
-	public function get_name() {
-		return $this->_data['name'] ? $this->_data['name'] : __( 'Fee', 'woocommerce' );
+	public function get_name( $context = 'view' ) {
+		$name = $this->get_prop( 'name', $context );
+		return $name ? $name : __( 'Fee', 'woocommerce' );
```
Comment (mikejolley) [inline includes/class-wc-order-item-fee.php]: This should either be handled on read, or only when in view context?

File: includes/class-wc-order-item-product.php [hunk 2]
```diff
@@ -391,73 +219,235 @@ public function get_type() {
...
+	 * Get any associated downloadable files.
+	 *
+	 * @return array
+	 */
+	public function get_item_downloads() {
+		$files   = array();
+		$product = $this->get_product();
+		$order   = $this->get_order();
+		if ( $product && $order && $product->is_downloadable() && $order->is_download_permitted() ) {
+			$download_ids = $this->data_store->get_download_ids( $this, $order );
```
Comment (mikejolley) [inline includes/class-wc-order-item-product.php]: Might be a change after order data PR is in - I added a custom download data store.

**Review by mikejolley — COMMENTED** [review]
Added a few items. Looks good overall - lets merge after the order data PR? Some things might clash (downloads).

**Comment by justinshreve** [comment]
>  lets merge after the order data PR?

Sounds good -- I'll update the download IDs part once your PR is in.

**Comment by justinshreve** [comment]
Finished remaining items, rebased, and re-tested.

File: includes/api/legacy/v3/class-wc-api-orders.php [hunk 3]
```diff
@@ -954,35 +954,37 @@ protected function set_line_item( $order, $item, $action ) {
...
-			$item = new WC_Order_Item_Product();
+			$line_item = new WC_Order_Item_Product();
 		} else {
-			$item = new WC_Order_Item_Product( $item['id'] );
+			$line_item = new WC_Order_Item_Product( $item['id'] );
 		}
 
-		$item->set_product( $product );
-		$item->set_order_id( $order->get_id() );
+		error_log( print_r ( $item, 1 ) );
```
Comment (mikejolley) [inline includes/api/legacy/v3/class-wc-api-orders.php]: Error log

**Review by mikejolley — CHANGES_REQUESTED** [review]
I think this is good to merge once the error log is removed :)

**Pull request merged**